### PR TITLE
HTTP API account/balance handling invalid addresses

### PR DESCRIPTION
**Pull request by velzevur**
[Pivotal task](https://www.pivotaltracker.com/story/show/153214980)

### CODE REVIEW

**Comment by velzevur** [comment]
I couldn't chose between them. Suggestions are welcome

**Comment by hanssv** [comment]
I would go for "Invalid address" in this case

**Comment by velzevur** [comment]
Swagger description updated to have a consistent message `Invalid address`

File: apps/aehttp/src/aehttp_dispatch_ext.erl [hunk 1]
```diff
@@ -129,16 +129,23 @@ handle_request('GetAccountBalance', Req, _Context) ->
               {ok, PK} = aec_keys:pubkey(),
               PK;
           PK ->
-              base64:decode(PK)
+              try base64:decode(PK)
+              catch error:{badarg, _} -> not_base64_encoded
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_dispatch_ext.erl]: Ok.

----

I tested locally:
```
4> f(), try base64:decode(<<"foo">>) catch E:R -> {E, R} end.
{error,{badmatch,false}}
```

File: apps/aehttp/src/aehttp_dispatch_ext.erl [hunk 2]
```diff
@@ -129,16 +129,23 @@ handle_request('GetAccountBalance', Req, _Context) ->
...
-    {ok, LastBlock} = aec_chain:top(),
-    Trees = aec_blocks:trees(LastBlock),
-    AccountsTree = aec_trees:accounts(Trees),
-    case aec_accounts:get(Pubkey, AccountsTree) of
-        {ok, #account{balance = B}} ->
-            {200, [], #{balance => B}};
+    case Pubkey of
+        not_base64_encoded ->
+            {400, [], #{reason => <<"Invalid address">>}};
         _ ->
```
Comment (lucafavatella) [inline apps/aehttp/src/aehttp_dispatch_ext.erl]: This open case is fragile as the atom of the clause above may be mistyped by mistake. I suggest adding guard (e.g. `when is_binary(...)`).

Comment (velzevur) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: One can never be too assertive, I've pushed a change for this

File: apps/aehttp/src/aehttp_dispatch_ext.erl (hunk 1 above)
Comment (lucafavatella) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: (Just to be clear: no action on this comment - this comment states that this line looks correct.)

Comment (lucafavatella) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: I correct myself: there is issue. base64:decode returns two kinds of errors: badmatch and badarg.

Comment (lucafavatella) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]:
```
6> f(), try base64:decode(<<"bob_address">>) catch E:R -> {E,R} end.
{error,{badarg,95}}
7> f(), try base64:decode(<<"bobaddress">>) catch E:R -> {E,R} end.
{error,{badmatch,false}}
```

Comment (lucafavatella) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: So in one case 400 is returned, in other case 500 is returned.
```
$ curl -v http://localhost:3013/v1/account/balance?pub_key=bobaddress
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3013 (#0)
> GET /v1/account/balance?pub_key=bobaddress HTTP/1.1
> Host: localhost:3013
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 500 Internal Server Error
< server: Cowboy
< date: Wed, 29 Nov 2017 12:09:18 GMT
< content-length: 0
< content-type: application/json
<
* Connection #0 to host localhost left intact

$ curl -v http://localhost:3013/v1/account/balance?pub_key=bob_address
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3013 (#0)
> GET /v1/account/balance?pub_key=bob_address HTTP/1.1
> Host: localhost:3013
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 400 Bad Request
< server: Cowboy
< date: Wed, 29 Nov 2017 12:09:21 GMT
< content-length: 28
< content-type: application/json
<
* Connection #0 to host localhost left intact
```

Comment (lucafavatella) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: So this PR catches the case when the pubkey specified contains chars not in the base64 range. Though does not catch the case that the chars are in the base64 range though it cannot be decoded.

File: apps/aehttp/src/aehttp_dispatch_ext.erl [hunk 3]
```diff
@@ -128,17 +128,24 @@ handle_request('GetAccountBalance', Req, _Context) ->
           undefined ->
               {ok, PK} = aec_keys:pubkey(),
               PK;
-          PK ->
-              base64:decode(PK)
+          PK when is_binary(PK) ->
+              try base64:decode(PK)
+              catch _: _ -> not_base64_encoded
```
Comment (hanssv) [inline apps/aehttp/src/aehttp_dispatch_ext.erl]: ugly space after ':'

Comment (velzevur) [reply apps/aehttp/src/aehttp_dispatch_ext.erl]: ugly space is no more

**Comment by velzevur** [comment]
@lucafavatella does this account has a balance? We create accounts when they actually receive a balance so while there is no coinbase reward - this account is not created.

**Comment by lucafavatella** [comment]
> does this account has a balance? We create accounts when they actually receive a balance so while there is no coinbase reward - this account is not created.

Yeah - I just figured out locally.

**Comment by lucafavatella** [comment]
I see positive case works.

I set mining autostart, then:
```
$ curl http://localhost:3113/v1/account/pub-key
{"pub_key":"BMx8Lfj42Q7vkESMdF8a0Z01JSc/jwyUGzfPcwZYN9zO/X+8fV6CAmm1Od0OqxWq7SFYv/astoPFGWUzNyPvL8I="}

$ curl -v -G http://localhost:3013/v1/account/balance --data-urlencode "pub_key=BMx8Lfj42Q7vkESMdF8a0Z01JSc/jwyUGzfPcwZYN9zO/X+8fV6CAmm1Od0OqxWq7SFYv/astoPFGWUzNyPvL8I="
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 3013 (#0)
> GET /v1/account/balance?pub_key=BMx8Lfj42Q7vkESMdF8a0Z01JSc%2FjwyUGzfPcwZYN9zO%2FX%2B8fV6CAmm1Od0OqxWq7SFYv%2FastoPFGWUzNyPvL8I%3D HTTP/1.1
> Host: localhost:3013
> User-Agent: curl/7.43.0
> Accept: */*
>
< HTTP/1.1 200 OK
< server: Cowboy
< date: Wed, 29 Nov 2017 13:48:52 GMT
< content-length: 15
< content-type: application/json
<
* Connection #0 to host localhost left intact
{"balance":100}
```

**Pull request merged**